src/sync/pull.rs

```diff
@@ -59,8 +60,18 @@ pub async fn begin_pull(
     let pull_timer = rlog::Timer::new().map_err(InternalTimerError)?;
     let pull_resp = puller
         .pull(&pull_req, &pull_url, &pull_auth, &request_id)
-        .await
-        .map_err(PullFailed)?;
+        .await;
+    if let Err(PullError::FetchNotOk(status_code, body)) = pull_resp {
```

Comment:
Yeah, looking now at the code your suggestion to have puller return its httpinfo sounds like the way to go. The puller should pull and the caller can decide what to do with the info and response.